File: apps/ios/Exponential/UI/Agent/AgentActionPickerSheet.swift
import ExpCore
import ExpUI
import SwiftUI

/// EXP-825: the composer's action picker — the ▶ tool's single-pick list:
/// builtins pinned FIRST by the `builtin` flag ("Fix merge conflicts", then
/// "Create action" — creation is a subject of this composer now, not a sheet
/// of its own), then the team's rows (`AgentComposerModel.actions` owns that
/// order). Chat is never listed: it is what "no subject" means. A tap picks
/// and dismisses; picking swaps out any issue chips.
///
/// EXP-1030: it renders through the SHARED `ActionPicker` — every row the
/// action's curated glyph (EXP-273) + its name + its description, in the one
/// sheet every other pick opens. HOST-DRIVEN (`open`): the composer's tool
/// button is the trigger, in another view tree.
struct AgentActionPickerSheet: View {
    let model: AgentComposerModel
    /// The composer's ▶ button drives it.
    @Binding var isPresented: Bool

    var body: some View {
        ActionPicker(
            // EXP-273: the action's own curated glyph (the builtins set one
            // too), falling back to the generic action mark — resolved by the
            // ONE row bridge, `ActionPickerAction(ActionDto)`.
            actions: model.actions.map(ActionPickerAction.init),
            value: model.selectedAction?.id,
            onChange: { id in
                guard let action = model.actions.first(where: { $0.id == id }) else { return }
                model.pickAction(action)
            },
            open: $isPresented,
            hideTrigger: true,
            sheetIdentifier: "agent-composer-actions-picker",
            trigger: { EmptyView() }
        )
    }
}

File: apps/android/app/src/main/java/com/exponential/app/ui/agent/AgentActionPickerSheet.kt
package com.exponential.app.ui.agent

import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import com.exponential.app.data.api.ActionDto
import com.exponential.app.ui.components.picker.ActionPicker
import com.exponential.app.ui.components.picker.ActionPickerAction

/**
 * EXP-825: the composer's ▶ tool — a SINGLE-select action list (the
 * Start-coding sheet's Actions tab, EXP-257): "Fix merge conflicts" first,
 * then "Create action" (listed here now that the create sheet is gone — the
 * creator run is just another action to pick), then the team's rows in server
 * order; the hidden Chat row never (it is what "no subject" means). A tap
 * picks and closes; picking an action while issues are chipped SWAPS them out.
 *
 * EXP-1030: the sheet IS the shared [ActionPicker] — the curated glyph, the
 * name and the description, searched and marked the way every other pick on
 * this phone is. The loading and error states ride the primitive's ONE empty
 * text: with no row to offer, what the sheet has to say is why.
 */
@Composable
internal fun AgentActionPickerSheet(
    actions: List<ActionDto>?,
    error: String?,
    selectedId: String?,
    onPick: (String) -> Unit,
    onDismiss: () -> Unit,
) {
    ActionPicker(
        actions = actions.orEmpty().map { action ->
            ActionPickerAction(
                id = action.id,
                name = action.name,
                icon = action.icon,
                description = action.description?.takeIf { it.isNotBlank() },
            )
        },
        value = selectedId,
        // A single pick closes the primitive's sheet, which is what dismisses
        // this one (see `onOpenChange`) — the pick itself only chips it.
        onChange = onPick,
        title = "Actions",
        emptyText = when {
            actions == null && error != null -> error
            actions == null -> "Loading actions…"
            else -> "No matching actions."
        },
        sheetModifier = Modifier.testTag("agent-composer-actions-picker"),
        open = true,
        onOpenChange = { open -> if (!open) onDismiss() },
    )
}
